Code/wwlib/FFmpegFile.cpp

```diff
@@ -0,0 +1,379 @@
+		char error_buffer[1024];
+		av_strerror(result, error_buffer, sizeof(error_buffer));
+		WWDEBUG_SAY(("Failed 'avformat_find_stream_info': %s", error_buffer));
+		Close();
+		return false;
+	}
+
+	Streams.resize(FmtCtx->nb_streams);
+	for (unsigned int stream_idx = 0; stream_idx < FmtCtx->nb_streams; stream_idx++) {
+		AVStream *av_stream = FmtCtx->streams[stream_idx];
+		const AVCodec *input_codec = avcodec_find_decoder(av_stream->codecpar->codec_id);
+		if (input_codec == nullptr) {
+			WWDEBUG_SAY(("Codec not supported: '%s'", avcodec_get_name(av_stream->codecpar->codec_id)));
+			Close();
+			return false;
+		}
+
+		AVCodecContext *codec_ctx = avcodec_alloc_context3(input_codec);
+		if (codec_ctx == nullptr) {
+			WWDEBUG_SAY(("Could not allocate codec context"));
+			Close();
+			return false;
+		}
+
+		result = avcodec_parameters_to_context(codec_ctx, av_stream->codecpar);
+		if (result < 0) {
+			char error_buffer[1024];
+			av_strerror(result, error_buffer, sizeof(error_buffer));
+			WWDEBUG_SAY(("Failed 'avcodec_parameters_to_context': %s", error_buffer));
+			Close();
+			return false;
+		}
+
+		result = avcodec_open2(codec_ctx, input_codec, nullptr);
+		if (result < 0) {
+			char error_buffer[1024];
+			av_strerror(result, error_buffer, sizeof(error_buffer));
+			WWDEBUG_SAY(("Failed 'avcodec_open2': %s", error_buffer));
+			Close();
+			return false;
+		}
+
+		FFmpegStream &output_stream = Streams[stream_idx];
+		output_stream.codec_ctx = codec_ctx;
+		output_stream.codec = input_codec;
+		output_stream.stream_type = input_codec->type;
+		output_stream.stream_idx = stream_idx;
+		output_stream.frame = av_frame_alloc();
+	}
+
+	Packet = av_packet_alloc();
+
+	return true;
+}
+
+/**
+ * Read an FFmpeg packet from file
+ */
+int FFmpegFile::Read_Packet(void *opaque, uint8_t *buf, int buf_size)
+{
+	FileClass *file = static_cast<FileClass *>(opaque);
+	const int read = file->Read(buf, buf_size);
+
+	// Streaming protocol requires us to return real errors - when we read less equal 0 we're at EOF
+	if (read <= 0)
+		return AVERROR_EOF;
+
+	return read;
+}
+
+/**
+ * close all the open FFmpeg handles for an open file.
+ */
+void FFmpegFile::Close()
+{
+	if (FmtCtx != nullptr) {
+		avformat_close_input(&FmtCtx);
+	}
+
+	for (auto &stream : Streams) {
+		if (stream.codec_ctx != nullptr) {
+			avcodec_free_context(&stream.codec_ctx);
+			av_frame_free(&stream.frame);
+		}
+	}
+	Streams.clear();
+
+	if (AvioCtx != nullptr) {
+		av_freep(&AvioCtx->buffer);
+		avio_context_free(&AvioCtx);
+	}
+
+	if (Packet != nullptr) {
+		av_packet_free(&Packet);
+	}
+
+	if (File != nullptr) {
+		Factory->Return_File(File);
+		File = nullptr;
+	}
+
+	if (Factory != nullptr) {
+		Factory= nullptr;
+	}
+}
+
+bool FFmpegFile::Decode_Packet()
+{
+	WWASSERT_PRINT(FmtCtx != nullptr, ("null format context"));
+	WWASSERT_PRINT(Packet != nullptr, ("null packet pointer"));
+
+	int result = av_read_frame(FmtCtx, Packet);
+	if (result == AVERROR_EOF)
+		return false;
```

Comment:
I think the convention is to always use curly braces. (There are more occurrences of this)
```suggestion
	if (result == AVERROR_EOF) {
		return false;
	}
```